go/arrow/avro/reader_types.go

```diff
@@ -0,0 +1,892 @@
+	case map[string]any:
+		switch dt["int"].(type) {
+		case nil:
+			b.AppendNull()
+		case int32:
+			b.Append(arrow.Date32(dt["int"].(int32)))
+		}
+	}
+}
+
+func appendDecimal128Data(b *array.Decimal128Builder, data interface{}) error {
+	switch dt := data.(type) {
+	case nil:
+		b.AppendNull()
+	case []byte:
+		if len(dt) <= 38 {
+			var intData int64
+			buf := bytes.NewBuffer(dt)
+			err := binary.Read(buf, binary.BigEndian, &intData)
+			if err != nil {
+				return err
+			}
+			b.Append(decimal128.FromI64(intData))
+		} else {
+			var bigIntData big.Int
+			buf := bytes.NewBuffer(dt)
+			err := binary.Read(buf, binary.BigEndian, &bigIntData)
+			if err != nil {
+				return err
+			}
+			b.Append(decimal128.FromBigInt(&bigIntData))
+		}
+	case map[string]any:
+		if len(dt["bytes"].([]byte)) <= 38 {
+			var intData int64
+			buf := bytes.NewBuffer(dt["bytes"].([]byte))
+			err := binary.Read(buf, binary.BigEndian, &intData)
+			if err != nil {
+				return err
+			}
+			b.Append(decimal128.FromI64(intData))
+		} else {
+			var bigIntData big.Int
+			buf := bytes.NewBuffer(dt["bytes"].([]byte))
+			err := binary.Read(buf, binary.BigEndian, &bigIntData)
+			if err != nil {
+				return err
+			}
+			b.Append(decimal128.FromBigInt(&bigIntData))
+		}
+	}
+	return nil
+}
+
+func appendDecimal256Data(b *array.Decimal256Builder, data interface{}) error {
+	switch dt := data.(type) {
+	case nil:
+		b.AppendNull()
+	case []byte:
+		var bigIntData big.Int
+		buf := bytes.NewBuffer(dt)
+		err := binary.Read(buf, binary.BigEndian, &bigIntData)
+		if err != nil {
+			return err
+		}
+		b.Append(decimal256.FromBigInt(&bigIntData))
+	case map[string]any:
+		var bigIntData big.Int
+		buf := bytes.NewBuffer(dt["bytes"].([]byte))
+		err := binary.Read(buf, binary.BigEndian, &bigIntData)
+		if err != nil {
+			return err
+		}
+		b.Append(decimal256.FromBigInt(&bigIntData))
+	}
+	return nil
+}
+
+func appendDurationData(b *array.MonthDayNanoIntervalBuilder, data interface{}) {
+	switch dt := data.(type) {
+	case nil:
+		b.AppendNull()
+	case []byte:
+		dur := new(arrow.MonthDayNanoInterval)
+		dur.Months = int32(binary.BigEndian.Uint64(dt[:3]))
+		dur.Days = int32(binary.BigEndian.Uint64(dt[4:7]))
+		dur.Nanoseconds = int64(binary.BigEndian.Uint64(dt[8:]) * 1000000)
+		b.Append(*dur)
+	case map[string]any:
+		switch dtb := dt["bytes"].(type) {
+		case nil:
+			b.AppendNull()
+		case []byte:
+			dur := new(arrow.MonthDayNanoInterval)
+			dur.Months = int32(binary.BigEndian.Uint64(dtb[:3]))
+			dur.Days = int32(binary.BigEndian.Uint64(dtb[4:7]))
+			dur.Nanoseconds = int64(binary.BigEndian.Uint64(dtb[8:]) * 1000000)
+			b.Append(*dur)
+		}
+	}
+}
+
+func appendFixedSizeBinaryData(b *array.FixedSizeBinaryBuilder, data interface{}) {
+	switch dt := data.(type) {
+	case nil:
+		b.AppendNull()
+	case []byte:
+		b.Append(dt)
+	case map[string]any:
+		switch dt["bytes"].(type) {
+		case nil:
+			b.AppendNull()
+		case []byte:
+			b.Append(dt["bytes"].([]byte))
+		}
```

Comment:
You don't need the second type assertion here, just assign properly in the type switch:

```go
switch v := dt["bytes"].(type) {
case nil:
    b.AppendNull()
case []byte:
    b.Append(v)
}
```

You can do this for all of the append functions below to save yourself some effort. Is there the possibility we need to handle a default case or something other than these? Should we error? Panic? or should we have a default case that appends null?

Reply:
I've made the changes to remove the second type assertion, will push in next commit.
hamba/avro either returns nil, or the typed value, or a map[string]any with a single key named after the type (I believe this is in the case of nullable union types) -  I don't believe there are other cases to handle.